### ISSUE TITLE
tokenizer pipelines in apache beam

### ISSUE THREAD

**Issue by njunge94** [huggingface/tokenizers#638]
Hello! I have been using the tokenizers library in python for about a week now, and I really like its modular approach and customization options.

There is a topic that I'd like some input on. I want to port the tokenization process to another pipelining tool, in this case Apache Beam. This works fairly well for things like vocab computation, normalization and pre-tokenization, which I can handle by mapping these functions to my incoming stream of sentences. But the point that I'm struggling with is the `Model` object. 

As far as I could gather from looking at the source code (I have never worked with Rust), the actual word-to-token-list conversion inside the model (e.g. words to vocab/merges lists in the BPE Tokenizer) happens inside the `Tokenizer.train` method and is not exposed in a Python binding. So to achieve the above, I would probably have to move the Model algorithms to pure Python as well. Is that correct, or am I missing something here?

Thanks in advance for any comments!

**Comment by Narsil** [comment]
Hi @njunge94,

If you want to use tokenizers with Apache Beam, I encourage you to take a look at https://github.com/huggingface/datasets/ which we build for exactly that purpose. The documentation is here: https://huggingface.co/docs/datasets/quicktour.html.

As for you original question, yes the object in charge from changing strings into tokens (words is ill-defined in some languages). It happens in the `encode` function. `tokenizer.encode("This is a test").ids` should give you a list of ids of those tokens (the actual return has more information like offsets, actual tokens etc..). The `train` function you're referencing is a way to build a tokenizer from data and it *is* exposed to python. If you're building tokenizers from scratch, you should check our quicktour: https://huggingface.co/docs/tokenizers/python/latest/quicktour.html#build-a-tokenizer-from-scratch which explains how to access those functions.

Have I answered your question ?

**Comment by njunge94** [comment]
Hey @Narsil ,

thank you for the response! Yes, I have checked out `datasets` as well - in fact, I'm working on one of the available datasets just now. 

I think I made my question sound a little ambiguous, sorry about that. `tokenizer.encode`ing my tokens in the end is less of a problem as you rightfully pointed out, *once a trained tokenizer is available*. My question was directed at how I can `train` a tokenizer on a Beam pipeline of sentences as input instead of supplying filenames of text data to `tokenizer.train`.

My goal is to feed a Beam Pipeline of sentences in and train a huggingface tokenizer **on the pipelined data in Beam**, but the part that is stopping me is that there seems to be no easy way to pipeline the training process that happens inside of the `train` function in pure Python.I figured this was because of the speed and concurrency benefits of implementing this in Rust entirely, which makes sense.

**Comment by Narsil** [comment]
Oh perfect I got it, then you're probably looking for this:

https://huggingface.co/docs/tokenizers/python/latest/tutorials/python/training_from_memory.html#using-the-datasets-library

```python
import datasets                                                                     
from tokenizers import (                                                            
    Tokenizer,                                                                      
    models,                                                                         
    normalizers,                                                                    
    pre_tokenizers,                                                                 
    decoders,                                                                       
    trainers,                                                                       
)                                                                                   
                                                                                    
tokenizer = Tokenizer(models.Unigram())                                             
tokenizer.normalizer = normalizers.NFKC()                                           
tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel()                                
tokenizer.decoders = decoders.ByteLevel()                                           
                                                                                    
trainer = trainers.UnigramTrainer(                                                  
    vocab_size=20000,                                                               
    initial_alphabet=pre_tokenizers.ByteLevel.alphabet(),                           
    special_tokens=["<PAD>", "<BOS>", "<EOS>"],                                     
)                                                                                   
                                                                                    
                                                                                    
dataset = datasets.load_dataset(                                                    
    "wikitext", "wikitext-103-raw-v1", split="validation"                           
)                                                                                   
                                                                                    
                                                                                    
def batch_iterator(batch_size=1000):                                                
    for i in range(0, len(dataset), batch_size):                                    
        yield dataset[i: i + batch_size]["text"]                                    
                                                                                    
                                                                                    
tokenizer.train_from_iterator(                                                      
    batch_iterator(), trainer=trainer, length=len(dataset)                          
)                                                                                   
tokenizer.save("tokenizer.json")   
```

**Comment by njunge94** [comment]
Hey @Narsil ,

that looks pretty good! Thank you for taking the time to write this out. 

In the end, I decided to cut my losses. I am working with TensorFlow Extended (TFX), which forced me into some awkward decisions because of its lack of integrations with HuggingFace (not meant to be a knock on you guys, by the way). I now stash everything in my pipeline into intermediate cached text files, which I feed into my tokenizer the "intended" way.

From then on, like you said, it is a quick map of `tokenizer.encode` onto my input stream, plus truncation / padding. 

Thanks for the help! I'm closing this if you don't mind, since I made it work satisfactorily for me.

**Comment by lschneidpro** [comment]
> Hey @Narsil ,
> 
> that looks pretty good! Thank you for taking the time to write this out.
> 
> In the end, I decided to cut my losses. I am working with TensorFlow Extended (TFX), which forced me into some awkward decisions because of its lack of integrations with HuggingFace (not meant to be a knock on you guys, by the way). I now stash everything in my pipeline into intermediate cached text files, which I feed into my tokenizer the "intended" way.
> 
> From then on, like you said, it is a quick map of `tokenizer.encode` onto my input stream, plus truncation / padding.
> 
> Thanks for the help! I'm closing this if you don't mind, since I made it work satisfactorily for me.

Hi njunge94, I was wondering if you finally managed to make it work with TFX. I'm looking for an MLOps framework to deploy some transformers. Thanks

**Comment by njunge94** [comment]
> 
> Hi njunge94, I was wondering if you finally managed to make it work with TFX. I'm looking for an MLOps framework to deploy some transformers. Thanks

Hi! I worked on this during my time working on developing ZenML, an open-source MLOps framework based on TFX. It should be a good candidate for what you want. You can check it out here: https://github.com/maiot-io/zenml

There, you can find a demo example of a NLP Pipeline with HuggingFace components, which I made. You can run the example with the instructions you find there.

Best regards, 
Nicholas

**Comment by lschneidpro** [comment]
> > Hi njunge94, I was wondering if you finally managed to make it work with TFX. I'm looking for an MLOps framework to deploy some transformers. Thanks
> 
> Hi! I worked on this during my time working on developing ZenML, an open-source MLOps framework based on TFX. It should be a good candidate for what you want. You can check it out here: https://github.com/maiot-io/zenml
> 
> There, you can find a demo example of a NLP Pipeline with HuggingFace components, which I made. You can run the example with the instructions you find there.
> 
> Best regards,
> Nicholas

Hi Nicholas. thanks for your answer. I need to investigate both ZenML and TFX. My plan with TFX was to use to distilbert and use tensorflowhub's BERT tokenizer. Do you have any advice to give me so far based on your own experience? Thanks

**Comment by njunge94** [comment]
> Hi Nicholas. thanks for your answer. I need to investigate both ZenML and TFX. My plan with TFX was to use to distilbert and use tensorflowhub's BERT tokenizer. Do you have any advice to give me so far based on your own experience? Thanks

I think that HF's tokenizer implementation is fast enough for a ML pipeline setting. So if you do not want to get vendor-locked into TFX too much, you can easily adopt HF tokenizers in a more general pipeline (like you can see in the ZenML example).

Otherwise, if you are really fully attached to Tensorflow already, you can always go full TFX and grab the BERT models from TF's ModelHub, too. AFAIK, TFX has an end-to-end BERT pipeline doing just that.

From my experience, up until training, everything works fine with either solution. It is the subsequent model deployment where things start to get interesting :-) and there, my guess is that TFX has the edge (for now).

**Comment by tanzhenyu** [comment]
> https://github.com/huggingface/datasets/

I actually think this is a good candidate for TFX/Keras to support. Happy to initiate the discussion

**Comment by tanzhenyu** [comment]
> Hey @Narsil ,
> 
> that looks pretty good! Thank you for taking the time to write this out.
> 
> In the end, I decided to cut my losses. I am working with TensorFlow Extended (TFX), which forced me into some awkward decisions because of its lack of integrations with HuggingFace (not meant to be a knock on you guys, by the way). I now stash everything in my pipeline into intermediate cached text files, which I feed into my tokenizer the "intended" way.
> 
> From then on, like you said, it is a quick map of `tokenizer.encode` onto my input stream, plus truncation / padding.
> 
> Thanks for the help! I'm closing this if you don't mind, since I made it work satisfactorily for me.

I actually think this is a good candidate for TFX/Keras to support. Happy to initiate the discussion
